internal/core/command/restDevice.go

```diff
@@ -29,12 +30,68 @@ func restPutDeviceCommandByCommandID(w http.ResponseWriter, r *http.Request) {
 	issueDeviceCommand(w, r, true)
 }
 
+func restGetDeviceCommandByNames(w http.ResponseWriter, r *http.Request) {
+
+	//	Attempt to get command/device IDs to use
+	did, cid, err := convertDeviceCommandToIDs(w, r)
+	if err != nil {
+		LoggingClient.Error(err.Error())
+		return
+	}
+
+	issueDeviceCommandWork(w, r, did, cid, false)
+}
+
+func restPutDeviceCommandByNames(w http.ResponseWriter, r *http.Request) {
+	//	Attempt to get command/device IDs to use
+	did, cid, err := convertDeviceCommandToIDs(w, r)
+	if err != nil {
+		LoggingClient.Error(err.Error())
+		return
+	}
+
+	issueDeviceCommandWork(w, r, did, cid, true)
+}
+
+func convertDeviceCommandToIDs(w http.ResponseWriter, r *http.Request) (did string, cid string, err error) {
+
+	//	Setup
+	vars := mux.Vars(r)
+	dn := vars[NAME]
+	cn := vars[COMMAND]
+	ctx := r.Context()
+
+	//	Get Device
+	d, err := mdc.DeviceForName(dn, ctx)
+	if err != nil {
+		http.Error(w, "Device not found", http.StatusNotFound)
+		return "", "", errors.NewErrInvalidName(dn)
+	}
+
+	//	Match command name with commands associated with Device Profile.
+	for _, one := range d.Profile.Commands {
+		if cn == one.Name {
+			return d.Id, one.Id, nil;
+		}
+	}
+
+	//	Command name specified not found in commands of device's profile.
+	http.Error(w, "Command not found", http.StatusNotFound)
+	return "", "", errors.NewErrInvalidName(cn)
+}
+
 func issueDeviceCommand(w http.ResponseWriter, r *http.Request, p bool) {
-	defer r.Body.Close()
 
 	vars := mux.Vars(r)
 	did := vars[ID]
 	cid := vars[COMMANDID]
+	issueDeviceCommandWork(w, r, did, cid, p)
+}
+
+func issueDeviceCommandWork(w http.ResponseWriter, r *http.Request, did string, cid string, p bool) {
```

Comment:
It looks like this was created due to the variance in looking up devices/commands by name instead of ID. How else could you do this in a way that makes the intent obvious to later developers? `issueDeviceCommand` versus `issueDeviceCommandWork` doesn't really communicate a lot about why the logic is branched and at first I missed it.

Reply:
Following pattern used by REST methods by IDs, create two methods for REST methods by names, followed by the corresponding non-rest functions that extract the IDs or names, as appropriate.  Then the "work" method becomes plain-old issueDeviceCommand because that's all it does, no data grabbing or interpretation, get the IDs and act upon them.